Introducing Samsung Exynos i S111 target to Mbed-OS

andrewc-arm opened the pull request:
Original Issue Case: https://github.com/ARMmbed/mbed-os-rjio/issues/20

### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

Adding a new target of HW development kit using [Samsung Exynos i S111](https://www.samsung.com/semiconductor/minisite/exynos/products/iot/exynos-i-s111/) module to Mbed-OS. This will widen the HW choices of Mbed-OS enabled NB-IoT, GNSS and Security (eFuse, AES, SHA-2, PKA, Secure Storage, Security Sub-System, [PUF](https://en.wikipedia.org/wiki/Physical_unclonable_function)) modules.
Target Name: S5JS100

Code Managers:
    - JiHoon Park <jh6186.park@samsung.com>, github: jh6186
    - JongJin Lee <jjssm.lee@samsung.com>, github: jjssmlee

#### Compile Logs
[build_log_blinky_K64F_&_S5JS100.zip](https://github.com/ARMmbed/mbed-os/files/2987827/build_log_blinky_K64F_._S5JS100.zip)
[greentea_compile_log_S5JS100.zip](https://github.com/ARMmbed/mbed-os/files/3031962/greentea_compile_log_S5JS100.zip)

#### mbed-os-tool PR for supporting new target: S5JS100
https://github.com/ARMmbed/mbed-os-tools/pull/142

#### Greentea log
Greentea logs are being prepared and will be attached here, soon.

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

dannybenor commented on targets/targets.json:
```diff
@@ -5270,6 +5270,31 @@
+            "FLASH",
+            "STORAGE"
+        ],
+        "features": ["STORAGE", "SSS"],
```
Please remove "features": ["STORAGE". Not needed and deprecated

andrewc-arm replied:
Hi, @dannybenor. Thanks for the feedback. I just submitted 82a2a476a30891d0854712d785ee08c1c37e64f9.

davidsaada commented on features/security_subsystem/FEATURE_SSS/api/trng_api.c:
```diff
@@ -0,0 +1,40 @@
+/****************************************************************************
```
General note: This PR adds a new subsystem called "security_subsystem" with a new feature called "SSS". This is something that's used for general features, not for board specific ones. 
Now, in this case, it seems that the feature added here is TRNG. We have a very well defined interface for that under hal/trng_api.h. Board specific implementations for such features should be placed under target code. For reference, you can look at the many boards implementing trng under "target" directory (code that includes "trng_api.h").

andrewc-arm replied:
Addressed in the latest fc422106e140aaddc7ef5e7869d45a966e3c78cc

davidsaada commented on features/storage/FEATURE_STORAGE/cfstore/source/cfstore_config.h:
```diff
@@ -49,8 +49,11 @@
  *   Disable use of storage support (if present)
  */
 #if defined DEVICE_STORAGE && CFSTORE_STORAGE_DISABLE==0
+#ifdef TARGET_K64F
```
As @dannybenor mentioned above, cfstore is deprecated.

andrewc-arm replied:
Thanks. I fixed it on e578320f5aea90759f813e9a6f1eea2d2f884936.

davidsaada commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/BINARIES/bootarg_Board_0.2.bin:
```diff
@@ -0,0 +1 @@
+BOARD=0.2
```
Is this a binary file?

andrewc-arm replied:
Addressed from https://github.com/ARMmbed/mbed-os/pull/10170/files#diff-6ab48777b97d4bd41471af54d2b19faeR1

davidsaada commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/BINARIES/bootarg_default.bin:
```diff
@@ -0,0 +1 @@
+BOARD=0.4
```
Same here. and what's its purpose beside that?

andrewc-arm replied:
Addressed from https://github.com/ARMmbed/mbed-os/pull/10170/files#diff-6ab48777b97d4bd41471af54d2b19faeR1

davidsaada commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/BINARIES/partition.map:
```diff
@@ -0,0 +1,28 @@
+#
```
Why is a map file included here?

andrewc-arm replied:
Removed in the latest fc42210

davidsaada reviewed (changes requested):
Please see all my comments above, specially the ones regarding the new subsystem.
In addition:
- There are lots and lots of binaries in this PR. Are they all required? Are they all pre-built or can they be generated? If it's the latter, then they should be removed. Some of which look very suspicious, such as a "hello world" binary. In general, only very essential binaries should be included.
- Not sure that license headers follow the rules, but will leave this one for the maintainers.

davidsaada commented on targets/targets.json:
```diff
@@ -5270,6 +5270,31 @@
+            "SPI",
+            "SLEEP",
+            "FLASH",
+            "STORAGE"
```
Same here - should be removed.

andrewc-arm replied:
Thanks. I fixed it on e578320f5aea90759f813e9a6f1eea2d2f884936.

davidsaada commented on tools/config/__init__.py:
```diff
@@ -69,7 +69,7 @@
 
 ALLOWED_FEATURES = [
-    "BOOTLOADER", "BLE", "LWIP", "STORAGE", "NANOSTACK", "CRYPTOCELL310",
+    "BOOTLOADER", "BLE", "LWIP", "STORAGE", "NANOSTACK", "CRYPTOCELL310", "SSS",
```
Should be removed (see my comment above about board specific features and subsystems).

andrewc-arm replied:
Addressed in the latest fc42210

janjongboom committed:
Target should not define mbed_main, as it's user defined hook

andrewc-arm committed:
First code set of Samsung Exynos i S111 target to Mbed-OS.
Target Name: S5JS100
Code Managers:
    - JiHoon Park <jh6186.park@samsung.com>, github: jh6186
    - JongJin Lee <jjssm.lee@samsung.com>, github: jjssmlee

andrewc-arm committed:
Fixed compile errors during Greentea compile.

davidsaada commented on platform/mbed_rtc_time.h:
```diff
@@ -34,6 +34,8 @@ struct timeval {
     int32_t tv_usec;
 };
+#else
+#include <sys/time.h>
```
Why is this needed?

andrewc-arm replied:
Fixed in fc422106e140aaddc7ef5e7869d45a966e3c78cc

davidsaada commented on tools/export/iar/iar_definitions.json:
```diff
@@ -345,5 +345,8 @@
+        "OGChipSelectEditMenu": "S5JS100\tSamsung S5JS100"
     }
-}
+} 
```
Looks like a space has slipped in (sorry for nitpicking).

andrewc-arm replied:
Addressed from 4a063328bb2f02333932669e00fa7d7a512e215e

davidsaada reviewed (changes requested):
Thanks for the changes @andrewc-arm. This looks much better. Got a very minor comment and a question.
In addition, it looks like this PR interfaces the TRNG properly now. Did you run the `tests-mbed_hal-trng` test?

andrewc-arm committed:
Verified that this <sys/time.h> is not needed.

andrewc-arm committed:
Removing unnecessary space addition.

davidsaada reviewed (approved):
Thanks. Looks good now.

bridadan commented on targets/targets.json:
```diff
@@ -5306,6 +5306,27 @@
+        "inherits": ["Target"],
+        "core": "Cortex-M7",
+        "components_add": ["FLASHIAP"],
+        "supported_toolchains": ["GCC_ARM", "IAR", "ARM"],
```
Just because we had some recent changes regarding what `"ARM"` means in this field as of 5.12.0, does this target support ARMC5 and ARMC6?

andrewc-arm replied:
Hi,
As you know, all the recent Mbed-OS master based code compiles only in ARMC6. I couldn't test in ARMC5 because toolchain forced ARMC6 compilation. Here is the ARMC6 compile log: [build_log_S5JS100_ARM.txt](https://github.com/ARMmbed/mbed-os-rjio/files/3016168/build_log_S5JS100_ARM.txt)
```
/home/andcho01/tools/ARMCompiler6.12/bin/armclang -c --target=arm-arm-none-eabi -mthumb -Os -Wno-armcc-pragma-push-pop -Wno-armcc-pragma-anon-unions -DMULADDC_CANNOT_USE_R7 -fdata-sections -fno-exceptions -MMD -D_LIBCPP_EXTERN_TEMPLATE(...)= -fshort-enums -fshort-wchar -DMBED_TRAP_ERRORS_ENABLED=1 -mcpu=cortex-m7 -fno-rtti -std=gnu++98 -DMBED_MPU_CUSTOM -D__MBED__=1 -DTARGET_LIKE_MBED -DMBED_BUILD_TIMESTAMP=1553745562.35 -DDEVICE_PORTINOUT=1 -D__MBED_CMSIS_RTOS_CM -DCOMPONENT_PSA_SRV_EMUL=1 -D__CMSIS_RTOS -DTOOLCHAIN_ARMC6 -DARM_MATH_CM7 -DTARGET_CORTEX_M -DTARGET_LIKE_CORTEX_M7 -DTARGET_S5JS100 -DTARGET_M7 -DCOMPONENT_PSA_SRV_IMPL=1 -DCOMPONENT_FLASHIAP=1 -DCMSDK_CM7 -DDEVICE_INTERRUPTIN=1 -DTARGET_CORTEX -DDEVICE_PORTOUT=1 -D__CORTEX_M7 -DTARGET_Samsung -DDEVICE_PORTIN=1 -DTARGET_RELEASE -DCOMPONENT_NSPE=1 -DMBED_TEST_MODE -DTARGET_SIDK_S5JS100 -DDEVICE_TRNG=1 -DTARGET_RTOS_M4_M7 -DDEVICE_SLEEP=1 -DDEVICE_SPI=1 -DTARGET_NAME=S5JS100 -DDEVICE_SERIAL=1 -DDEVICE_FLASH=1 -DTOOLCHAIN_ARM -DTOOLCHAIN_ARM_STD @./BUILD/tests/S5JS100/ARM/.includes_700d7631b8f26d37481060463f56c96d.txt -include ./BUILD/tests/S5JS100/ARM/mbed_config.h -o BUILD/tests/S5JS100/ARM/mbed-os/features/netsocket/InternetSocket.o ./mbed-os/features/netsocket/InternetSocket.cpp
```

Also please note that I upgraded IAR as well to 8.32.2 as the latest Mbed-OS suggests.
[build_log_S5JS100_IAR.txt](https://github.com/ARMmbed/mbed-os-rjio/files/3016167/build_log_S5JS100_IAR.txt)

bridadan reviewed (commented):
One quick clarification that I've posted below.

The only real concern I have about this is the number of binaries and the size of `targets/TARGET_Samsung/TARGET_SIDK_S5JS100/BINARIES/cp_main_S111NB_SMDK_ONESKU_QB2029203.bin` in particular: 3.76 MB!

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/BINARIES/README.md:
```diff
@@ -0,0 +1,19 @@
+Explanation of files
+* pbl_dcache_jumpto_mbedos.bin: 
```
These binaries will need license files, reference https://os.mbed.com/docs/mbed-os/v5.12/contributing/license.html. It needs license files to be added

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/PeripheralNames.h:
```diff
@@ -0,0 +1,87 @@
+/****************************************************************************
+ *
+ * Copyright 2019 Samsung Electronics All Rights Reserved.
+ *
```
> Each source header should start with your copyright line, the SPDX identifier and the Apache 2.0 header as shown here:

please add SPDX identifier to all new files

0xc0170 commented on targets/TARGET_Samsung/security_subsystem/drivers/mb_cmd_rng.c:
```diff
@@ -0,0 +1,164 @@
+/**************************************************************************/
+/* Copyright (c) Samsung Electronics Co., Ltd. All rights reserved.       */
+/*                                                                        */
+/* -INSTRUCTIONS-                                                         */
```
This is not permissive license, please review https://os.mbed.com/docs/mbed-os/latest/contributing/license.html

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/serial_api.c:
```diff
@@ -0,0 +1,570 @@
+	UARTName uart_tx = (UARTName)pinmap_peripheral(tx, PinMap_UART_TX);
+	UARTName uart_rx = (UARTName)pinmap_peripheral(rx, PinMap_UART_RX);
+	UARTName uart = (UARTName)pinmap_merge(uart_tx, uart_rx);
+	if ((int)uart == NC) {
```
I've noticed coding style differences in HAL files like spi_api.c or serial_api.c - using 8 spaces and similar. Can you run astyle on these files please? it would be good to have HAL implementation following the coding style if we are adding new files

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/serial_api.c:
```diff
@@ -0,0 +1,570 @@
+		*(volatile unsigned int *)(0x83015018) = 0x0;
+		*(volatile unsigned int *)(0x8301501C) = 0x1;
+
+	/*  *(volatile unsigned int *)(0x83014020) = 0x0;  */
```
please remove the dead code

0xc0170 replied:
also on lines below, lines 209, 186 and others

Please check the rest of the files as well

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/modem/modem_io_device.h:
```diff
@@ -0,0 +1,159 @@
+#define __MODEM_IO_DEVICE_H__
+
+
+#include "mbed.h"
```
Please include rather specific headers instead of the application `mbed.h`

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/device/TOOLCHAIN_ARM_STD/sidk_s5js100.sct:
```diff
@@ -0,0 +1,40 @@
+#! armcc -E
+#define S5JS100_BOOTMEM_BASE    (0x00000000)
```
add a license header file  to this file as well

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/TOOLCHAIN_GCC_ARM/.gitattributes:
```diff
@@ -0,0 +1,4 @@
+*.o binary
```
dont we cover this in the gitattributes in the root already?

0xc0170 commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/TOOLCHAIN_GCC_ARM/.gitattributes:
```diff
@@ -0,0 +1,4 @@
+*.o binary
+*.lib binary
+*.a binary
```
I'll leave the comment here for libSAMSUNG_ libs. They need license files , following https://os.mbed.com/docs/mbed-os/latest/contributing/license.html

bulislaw commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/device/system_core_s5js100.c:
```diff
@@ -0,0 +1,327 @@
+	uint32_t *dest, size;
+	uint32_t vector;
+
+#if defined ( __ICCARM__ ) 
```
Mbed OS already copies the vector table to allow runtime changes: https://github.com/ARMmbed/mbed-os/blob/master/rtos/TARGET_CORTEX/mbed_boot.c#L104 can you review whether you can use the existing mechanism rather than redefining it. Also there's plenty of dead code here, please clean up.

bulislaw commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/port_api.c:
```diff
@@ -0,0 +1,79 @@
+}
+
+void port_init(port_t *obj, PortName port, int mask, PinDirection dir) {
+#if 0//revise me
```
Why are all the functions commented out?

bulislaw commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/s5js100_error.h:
```diff
@@ -0,0 +1,48 @@
+
+#ifdef S5JS100_ERROR_IS_HALT
+#define return_error(ret)		{ 																	\
+	if (S5JS100_ERROR_RETUN_LOG) printf("S5JS100 HALT!!! %s:%d:%s\n", __FILE__, __LINE__, __func__);	\
```
Could we avoid using printf? Mbed has debug calls that should be cheaper.

bulislaw commented on targets/TARGET_Samsung/TARGET_SIDK_S5JS100/sleep.c:
```diff
@@ -0,0 +1,138 @@
+	return 0;
+}
+
+static void s5js100_idle_hook(void)
```
Why are you redefining the idle hook? What's wrong with the default Mbed one? It's not meant for platform code to change, but for apps developers.
